rspace/src/test/scala/coop/rchain/rspace/nextgenrspace/ReplayRSpaceTests.scala

```diff
@@ -0,0 +1,995 @@
+      val channels     = List("ch1")
+      val patterns     = List(Wildcard)
+      val continuation = "continuation"
+      val datum        = "datum1"
+
+      for {
+        emptyPoint <- space.createCheckpoint()
+
+        resultConsume <- space.consume(channels, patterns, continuation, false)
+        resultProduce <- space.produce(channels(0), datum, false)
+        rigPoint      <- space.createCheckpoint()
+
+        _ = resultConsume shouldBe None
+        _ = resultProduce shouldBe Some(
+          (ContResult(continuation, false, channels, patterns, 1), List(Result(datum, false)))
+        )
+
+        _ <- replaySpace.rig(emptyPoint.root, rigPoint.log)
+
+        replayResultConsume <- replaySpace.consume(channels, patterns, continuation, false)
+        replayResultProduce <- replaySpace.produce(channels(0), datum, false)
+        finalPoint          <- space.createCheckpoint()
+
+        _ = replayResultConsume shouldBe None
+        _ = replayResultProduce shouldBe resultProduce
+        _ = finalPoint.root shouldBe rigPoint.root
+        _ = replaySpace.replayData shouldBe empty
+      } yield ()
+    }
+
+  implicit val arbitraryRangeSize: Gen[Int] = Gen.chooseNum[Int](1, 10)
+
+  "Picking a datum from n waiting datums" should "replay correctly" in forAll(arbitraryRangeSize) {
+    n: Int =>
+      fixture { (_, _, space, replaySpace) =>
+        for {
+          emptyPoint <- space.createCheckpoint()
+
+          range = (0 until n toList)
+          _ <- produceMany(
+                space,
+                range,
+                channelCreator = kp("ch1"),
+                datumCreator = i => s"datum$i",
+                persist = false
+              )
+          result <- space.consume(
+                     channels = List("ch1"),
+                     patterns = List(Wildcard),
+                     continuation = "continuation1",
+                     persist = false
+                   )
+          rigPoint <- space.createCheckpoint()
+          _        <- replaySpace.rig(emptyPoint.root, rigPoint.log)
+          _ <- produceMany(
+                replaySpace,
+                range,
+                channelCreator = kp("ch1"),
+                datumCreator = i => s"datum$i",
+                persist = false
+              )
+          replayResult <- replaySpace.consume(
+                           channels = List("ch1"),
+                           patterns = List(Wildcard),
+                           continuation = "continuation1",
+                           persist = false
+                         )
+          finalPoint <- replaySpace.createCheckpoint()
+          _          = replayResult shouldBe result
+          _          = replaySpace.replayData shouldBe empty
+          _          = finalPoint.root shouldBe rigPoint.root
+        } yield ()
+      }
+  }
+
+  "Picking n datums from n waiting datums" should "replay correctly" in forAll(arbitraryRangeSize) {
```

Comment:
why not have one test that picks n from m items instead of having two tests: one for an item out of n and one for n out of n?